app/src/main/java/com/quran/labs/androidquran/ui/PagerActivity.java

```diff
@@ -1837,6 +1887,217 @@ public void onError(@NonNull Throwable e) {
     );
   }
 
+  public void shareAyahAudio(SuraAyah start, SuraAyah end) {
+    audioCacheFilePaths.clear();
+
+    kotlin.Pair pair2 = getReorderedAyatPair(start, end);
+    selectedStartSuraAyah = (SuraAyah) pair2.component1();
+    selectedEndSuraAyah = (SuraAyah) pair2.component2();
+
+    selectedQari = audioStatusBar.getAudioInfo();
+    AudioPathInfo audioPathInfo = audioUtils.getLocalAudioPathInfo(selectedQari);
+
+    assert audioPathInfo != null;
+    boolean gaplessDatabaseExists = audioPathInfo.getGaplessDatabase() != null;
+
+    if (gaplessDatabaseExists) {
+      if (audioFilesExist(audioPathInfo)) {
+        createAndShareAudio(selectedStartSuraAyah, selectedEndSuraAyah, audioPathInfo);
+      } else {
+        requestDownload(audioPathInfo);
+      }
+    }
+  }
+
+  private kotlin.Pair getReorderedAyatPair(SuraAyah start, SuraAyah end) {
+    kotlin.Pair pair;
+    if (start.compareTo(end) <= 0) {
+      pair = TuplesKt.to(start, end);
+    } else {
+      Timber.Forest.e(
+          new IllegalStateException("End isn't larger than the start: " + start + " to " + end));
+      pair = TuplesKt.to(end, start);
+    }
+    return pair;
+  }
+
+  private boolean audioFilesExist(AudioPathInfo audioPathInfo) {
+    return audioUtils.haveAllFiles(audioPathInfo.getUrlFormat(), audioPathInfo.getLocalDirectory(),
+        selectedStartSuraAyah, selectedEndSuraAyah, true);
+  }
+
+  private void createAndShareAudio(SuraAyah start, SuraAyah end, AudioPathInfo audioPathInfo) {
+    showProgressDialog();
+    compositeDisposable.add(
```

Comment:
let's prefer coroutines to Rx - you'll be able to do that easier when we move out of this class

Reply:
Fixed in the new `AudioShareUtils` class in the `feature:audio`